File: solutions/problem_1982_find_array_given_subset_sums/solution.py
"""
Problem 1982: Find Array Given Subset Sums
========================================
Difficulty: Hard
Tags: Array, Divide and Conquer
Pattern: Divide and Conquer

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def recoverArray(self, n: int, sums: List[int]) -> List[int]:
        # Divide and conquer approach - O(n log n) time
        def solve(left, right):
            if left >= right:
                return n[left] if left < len(n) else 0
            mid = (left + right) // 2
            left_result = solve(left, mid)
            right_result = solve(mid + 1, right)
            return max(left_result, right_result)  # merge step
        
        return solve(0, len(n) - 1) if n else []

File: solutions/problem_1982_find_array_given_subset_sums/solution.cpp
/*
 * Problem 1982: Find Array Given Subset Sums
 * ========================================
 * Difficulty: Hard
 * Tags: Array, Divide and Conquer
 * Pattern: Divide and Conquer
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> recoverArray(int n, vector<int>& sums) {
        // Divide and conquer - O(n log n) time
        function<int(int, int)> solve = [&](int left, int right) -> int {
            if (left >= right) return left < (int)n.size() ? n[left] : 0;
            int mid = (left + right) / 2;
            int leftRes = solve(left, mid);
            int rightRes = solve(mid + 1, right);
            return max(leftRes, rightRes);
        };
        return n.empty() ? {} : solve(0, n.size() - 1);
    }
};
